File: src/logic/OneWordFiller.java
package logic;

import logic.declaration.Direction;
import logic.declaration.WordPosition;
import model.Cell;
import model.Crossword;

import java.util.List;

/**
 * User: Dmytro Vynokurov
 * Date: 10/6/13
 * Time: 12:12 PM
 */
public class OneWordFiller {

    public void matchAllWordsForCrossword(Crossword crossword, List<Crossword> newVariants) {
        WordPosition wordPosition = new FirstWordExtractor().extractStartWord(crossword);

        for(String word: crossword.getWords()){
            if (wordMatches(wordPosition,word)){
                newVariants.add(fillCrosswordWithSingleWord(crossword, wordPosition, word));
            }
        }
    }

    boolean wordMatches(WordPosition wordPosition, String word) {
        if (word.length() != wordPosition.getLength()) return false;

        Cell currentCell;
        for (int i = 0; i < word.length(); i++) {
            currentCell = wordPosition.getCell(i);
            if (currentCell == null) throw new IllegalStateException("Cell of a word cannot be null");
            if ((!currentCell.isEmpty()) && (currentCell.getValue() != word.charAt(i))) return false;
        }

        return true;
    }

    Crossword fillCrosswordWithSingleWord(Crossword crossword, WordPosition wordPosition, String word){
        Crossword copy = new Crossword(crossword,word);

        int currentRow = wordPosition.getStartRow();
        int currentColumn = wordPosition.getStartColumn();
        Direction direction = wordPosition.getDirection();
        Cell newCell;

        for(int i=0;i<wordPosition.getLength();i++){
            if(copy.getCell(currentRow,currentColumn).isEmpty()){
                newCell = new Cell(word.charAt(i));
                copy.setCell(newCell,currentRow,currentColumn);
            }
            currentRow = nextRow(currentRow, direction);
            currentColumn = nextColumn(currentColumn, direction);
        }
        return copy;
    }

    int nextColumn(int position, Direction direction){
        if(direction==Direction.HORIZONTAL) return position+1;
        else return position;
    }

    int nextRow(int position, Direction direction){
        if(direction==Direction.VERTICAL) return position+1;
        else return position;
    }
}

File: src/functional/logic/OneWordFiller.scala
package functional.logic

import model.{Cell, Crossword}
import logic.declaration.{Direction, WordPosition}
import FirstWordExtractor._

/**
 * User: Dmytro Vynokurov
 * Date: 12/8/13
 * Time: 11:37 PM
 */
object OneWordFiller {
  def matchAllWordsForCrossword(crossword:Crossword):List[Crossword] = {
    val wordPosition = extractStartWord(crossword)
    val firstLetterCoordinates = (wordPosition.getStartRow,wordPosition.getStartColumn)
    val dir = wordPosition.getDirection

    for(
      word <- crossword.getWords.toArray(new Array[String](0)).toList
      if wordMatches(wordPosition,word)
    ) yield {
      val variant = new Crossword(crossword,word)
      fillCrossWordWithSingleWord(variant,firstLetterCoordinates,dir,word.toCharArray.toList)
      variant
    }
  }
  
  private def wordMatches(wordPosition:WordPosition,word:String):Boolean = {
    if(wordPosition.getLength!=word.length)return false
    
    val cells = wordPosition.getCells.toList
    val letters = word.toCharArray.toList

    cells.zip(letters).forall((zip:(Cell,Char)) => {val (cell,char)=zip; cell.isEmpty || cell.getValue == char})
  }
  
  private def fillCrossWordWithSingleWord(crossword:Crossword,cell:(Int,Int),direction:Direction,letters:List[Char]):Unit = {
    if(!letters.isEmpty){
      val newCell = new Cell(letters.head)
      val (row,col) = cell
      crossword.setCell(newCell,row,col)

      fillCrossWordWithSingleWord(crossword,nextCell(cell,direction),direction,letters.tail)
    }
  }
  
  private def nextCell(a:(Int,Int),direction:Direction):(Int,Int) ={
    if(direction==Direction.VERTICAL) (a._1+1,a._2)
    else (a._1,a._2+1)
  }
}
